[RFC] Fix CudaEventCache for dangling references

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144301

Reported in https://github.com/pytorch/pytorch/issues/143470, we have a dangling references in CudaEventCache. So we want to fix it.
1. We add a unit test to repro the issue mentioned in the issue.
2. We tried to convert variables to shared pointers as suggested by @suo in the issue.


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

fduwjj committed:
Fix CudaEventCache for dangling references

[ghstack-poisoned]

c-p-i-o commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -438,6 +438,29 @@ def test_restart_pg(self):
+        side_thread = threading.Thread(target=init_collective_task, args=(t,))
+        side_thread.start()
+        side_thread.join()
+        torch.cuda.synchronize()
```
Is this test `asserting` something or it just runs to completion v/s crashing?

fduwjj replied:
with the current code, this test will crash. And after the fix, the test will pass so no assertion is needed. Actually we did check the return value in the unit test.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -806,16 +806,18 @@ ProcessGroupNCCL::CUDAEventCache::CUDAEventCache() = default;
-    std::lock_guard<std::mutex> lock(cacheMutex_);
-    auto& events = eventsArray_[timing ? 1 : 0];
+    std::lock_guard<std::mutex> lock(*cacheMutexPtr_);
+    auto& events = (*eventsArrayPtr_)[timing ? 1 : 0];
```
But this is not inside deleter.. let me think more on this.

suo replied:
oh, sorry I missed that

fduwjj replied:
Yeah the error is not related to this. And I get it working now.

fduwjj committed:
Update on "[RFC] Fix CudaEventCache for dangling references"


Reported in https://github.com/pytorch/pytorch/issues/143470, we have a dangling references in CudaEventCache. So we want to fix it.
1. We add a unit test to repro the issue mentioned in the issue.
2. We tried to convert variables to shared pointers as suggested by suo in the issue.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
This PR basically says: please don't destroy any of its members when a CUDAEventCache dies.
It may work; meanwhile it may complicate future extension as every member may need to follow this rule. 

I wonder if a slightly more extendable approach could be to not letting the cache die until references to it go to 0.
This may mean having a:
```
  static thread_local std::
      map<at::DeviceIndex, std::shared_ptr<ProcessGroupNCCL::CUDAEventCache>>
          cacheDeviceMap;
```
(Note the `std::shared_ptr` added here.)
Then hopefully the code inside `CUDAEventCache` can remain unchanged.

fduwjj commented:
@kwen2501 I submitted another PR for your proposed change: https://github.com/pytorch/pytorch/pull/144496. But I run into `double free or corruption (out)` issue. I need to think more on that one.

kwen2501 commented:
Thanks for trying it. I had a guess for the double free [here](https://github.com/pytorch/pytorch/pull/144496#discussion_r1910108229).

fduwjj committed:
Update on "[RFC] Fix CudaEventCache for dangling references"


Reported in https://github.com/pytorch/pytorch/issues/143470, we have a dangling references in CudaEventCache. So we want to fix it.
1. We add a unit test to repro the issue mentioned in the issue.
2. We tried to convert variables to shared pointers as suggested by suo in the issue.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]
